src/shared/sol-conffile.c

```diff
@@ -432,28 +463,112 @@ _fill_vector(const char *path, const char *fallback_paths)
     if (_json_to_vector(json_scanner) != 0)
         goto free_for_all;
 
-    if (include || include_fallbacks)
-        _fill_vector(include, include_fallbacks);
+    if (include || include_fallbacks.len) {
+        _fill_vector(include, &include_fallbacks);
+    }
 
 free_for_all:
     sol_file_reader_close(file_reader);
     free(full_path);
     free(include);
-    free(include_fallbacks);
+
+    SOL_VECTOR_FOREACH_IDX (&include_fallbacks, slice, i) {
+        free((char *)slice->data);
+    }
+
+    sol_vector_clear(&include_fallbacks);
+}
+
+static void
+_add_formated_lookup_path(struct sol_vector *vector, const char *fmt, ...)
+{
+    va_list ap;
+    char *buff;
+    int r;
+    struct sol_str_slice *slice;
+
+    va_start(ap, fmt);
+    r = vasprintf(&buff, fmt, ap);
+    va_end(ap);
+    if (r < 0 || !buff)
+        return;
+
+    slice = _vector_append_string_as_str_slice(vector, buff);
+    if (!slice) {
+        SOL_WRN("Couldn't append file path to vector.");
+        free(buff);
+    }
+}
+
+static void
+_add_lookup_path(struct sol_vector *vector, const char **search_dirs, char *appname, const char *board_name)
+{
+    size_t i;
+    struct sol_vector files = SOL_VECTOR_INIT(struct sol_str_slice);
+
+    if (appname && board_name) {
+        _add_formated_lookup_path(&files, "sol-flow-%s-%s.json", appname, board_name);
+    }
+
+    if (appname) {
+        _add_formated_lookup_path(&files, "sol-flow-%s.json", appname);
+    }
+
+    if (board_name) {
+        _add_formated_lookup_path(&files, "sol-flow-%s.json", board_name);
+    }
+
+    _add_formated_lookup_path(&files, "sol-flow.json");
+
+    for (i = 0; i < ARRAY_SIZE(search_dirs); i++) {
```

Comment:
Nope. In fact, it might be a good idea to change `ARRAY_SIZE` so that it checks if it's being applied to an array only. I believe kmod and the kernel have similar macros with these checks to serve as inspiration.